### PR TITLE
DOCS: Update development instructions

### PR DESCRIPTION
**Pull request by chrisjsewell**
- Link to the EBP development conventions (as suggested in https://github.com/executablebooks/meta/issues/79)
- Move/update release instructions to README.md. RELEASE.md feels like it should have contained the actual changelog, rather than just instructions for a release

@choldgraf do you think this practice of adding/removing `dev0` is important and should be part of the development conventions. I know you also do it in jupyter-book, but I've never done it myst-parser, markdown-it-py, etc, because I'm too lazy to deal with the added hassle lol.

### CODE REVIEW

**Comment by choldgraf** [comment]
This sounds great to me 👍 

re: `dev0`...I dunno, I've just always done it because it seems to be what folks recommend. That said, it annoys me and I think it's easier to follow the practices that you use 🤷

**Comment by chrisjsewell** [comment]
> That said, it annoys me and I think it's easier to follow the practices that you use

Yeh, it just feels that you should only be using the repo directly for development (i.e. `pip install -e .`), at which point I already know it may contain additional commits to the actual version specified; I don't need `dev0` to "tell" me that.

**Comment by choldgraf** [comment]
so then the general approach is:

* develop on `M.m.p` version
* make a release commit `RLS: M.m.p`
* bump to `M.m.p+1`
* tag the `RLS: ` commit for a release.

right?

**Comment by chrisjsewell** [comment]
> so then the general approach is:
>
>    develop on M.m.p version
>    make a release commit RLS: M.m.p
>    bump to M.m.p+1
>    tag the RLS: commit for a release.

Well, by the conventions `RLS` should be `🚀 RELEASE:` 😄 We can change this to `🚀 RLS:` if you want?

To clarify the order here: you want to tag/release the second from last commit?
How does "make a release commit" work if there is no actual changes to the repo, i.e. your not actually bumping a version here? Can you commit a null change?
I guess the only issue with working with a "bump to M.m.p+1" is that at this point you don't actually know whether the bump should be to `M`, `m` or `p`

Note in parallel with this I've made https://github.com/executablebooks/meta/pull/80

**Comment by choldgraf** [comment]
The reason that I make a release commit first, and then bump the version, just to ensure that there is always a canonical "released" version of the codebase. If you bump first, and then release, then you have a version of the code that's in the tarball of the release, and another version of the code with the same `__version__` that's on github `master` etc. Does that make sense? Again I don't feel *super* strongly about this, it's just the reasoning I've followed

**Comment by chrisjsewell** [comment]
Ermm, I can't really see myself using that convention:

I just do:

- make a release commit `🚀 RELEASE: ` which bumps `__version__` to `M.m.p`
- release/tag that commit as `vM.m.p`
- develop on `__version__ = "M.m.p"`

Then when I'm developing (`pip install -e .`) I "understand" that although the `__version__` is `M.m.p`, its actually "`M.m.p` + zero or more new commits".  I've never personally had an issue with this, or had any contributor express any confusion over it 🤷 
I feel `dev0` or having M.m.p one step ahead of the actual released version is only really necessary if your actually asking people to `pip install` directly from the github repo. Otherwise, if people want to work with a commit later than the current release in *production* (not to develop on), they should be requesting a pre-release to be made (e.g. `M.m.p-b0`).

Basically every time you change the version IMO it should relate directly to a specific tag, otherwise its kind of pointless

**Comment by choldgraf** [comment]
Yeah - I agree that's a simpler, in some ways more intuitive approach. I guess it's a question of whether to "break with convention" on this one.

I don't think `semver` explicitly calls out `dev0`, the closest thing to it is [this list about comparing version numbers](https://semver.org/#spec-item-11). I think if you have a "dev version" with one version number, and a "released" version with the exact same number, you couldn't use the logic of semver to decide which takes precedence (you'd have to look at the commit dates or something).

However, the [Python packaging guidelines](https://packaging.python.org/guides/distributing-packages-using-setuptools/?highlight=versioning#pre-release-versioning) do suggest a `dev0` versioning.

**Comment by chrisjsewell** [comment]
> However, the Python packaging guidelines do suggest a dev0 versioning.

But then it gives as an example versioning directly tied to a single commit:
`1.2.0.dev1+hg.5.b11e5e6f0b0b  # 5th VCS commmit since 1.2.0.dev1 release`,
as opposed to `dev0` which can refer to any commit post the last release.

I just think if your developing then you are working within a git repository and can use that to tell you the relationship of the code to the latest tag/release

**Comment by choldgraf** [comment]
> But then it gives as an example versioning directly tied to a single commit

well yeah but we're not suggesting putting the commit hash in the version either 😅 - I think the example they give makes sense, you could just assume "whatever is the latest commit for a given version takes precedence". What confuses me about bumping the version *before* the release is that the "latest commit" doesn't actually take precedence, instead it is "whatever commit is associated with a particular git tag".

OK now that I've said that - let's just go with the approach that you're describing, so long as we explain clearly in the developer instructions how it works. In the future if this creates some kind of unforeseen problem then we can always revisit the practice. That work?

**Comment by chrisjsewell** [comment]
I've exhausted you in to submission 🤣

**Comment by choldgraf** [comment]
yeah basically lol

**Comment by choldgraf** [comment]
(if you wanna update the release instructions here I'll merge afterward)

**Comment by chrisjsewell** [comment]
Ok I've added release instructions in executablebooks/meta@fc07c3ec2650a33129d998025ddc83c5e6de6c9f
If you want to review that PR first, then I'll merge that, alter the instructions here, and enact them for `v0.3.0`
